### PR TITLE
Add an endpoint selector to TestPlayground

### PR DESCRIPTION
**Pull request by wooj-stripe**
## Summary
Added a way to select your endpoint by storing a list of endpoints on the -v6 endpoint:
```
app.get("/endpoints", async (req, res) => {
  let endpointMap = {
    "LPM Betas": [
      "https://pool-seen-sandal.glitch.me/checkout",
      "https://some-other-endpoint.glitch.me/checkout"
    ],
    "Shared Endpoints": [
      "https://stripe-mobile-payment-sheet-test-playground-v6.glitch.me/checkout"      
    ]
  };
  let payload = {
    "endpoint_map" : endpointMap
  }
  res.send(payload);
});
```

## Motivation
Ability to select different endpoints for dogfooding/betas/etc.

## Testing
- Emulator: Fresh install, ensuring defaults are set/loaded
- Emulator: Subsequent installs
- Reseting all defaults, etc.

endpoint_selector

Annotation:
1.  Switches to GBP/GB merchant
2.  Loads backend for -v6 endpoint and creates new PI
3.  Loads payment sheet and shows two payment methods (card and paypal)
4.  Cancels payment sheet
5. Clicks on new "Endpoint" settings, which presents a modal
6.  Notice that the first section shows what the current endpoint
7.  User selects a different endpoint under section "LPM Betas", which applies these settings and dismisses the modal, and reloads automatically
8. payment sheet is presented, and since we are hitting the new endpoint which has 3 payment methods (card, paypal, and afterpay).  Then we dismiss payment sheet...
9. Finally, demonstrates that users can set the endpoint manually if they want.

### CODE REVIEW

File: Example/PaymentSheet Example/PaymentSheet Example/EndpointSelectorViewController.swift [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+extension EndpointSelectorViewController {
+    func loadEndpointSelector(endpoint: String) {
+        let request = URLRequest(url: URL(string: endpoint)!)
+        let session = URLSession.shared.dataTask(with: request) { data, response, error in
+            guard error == nil,
+                  let data = data,
+                  let specs = self.deserializeResponse(data: data) else {
+                print("failed")
+                return
+            }
```
Comment (wooj-stripe) [inline Example/PaymentSheet Example/PaymentSheet Example/EndpointSelectorViewController.swift:135]: Hmm... we should probably just bail at this point, and kick users back to the original screen if this happens.

Comment (wooj-stripe) [reply Example/PaymentSheet Example/PaymentSheet Example/EndpointSelectorViewController.swift:135]: Actually -- nevermind, we can't do that.  If people enter something manually, and the link is broken, they will never be able to tap "Set Manually"

File: Example/PaymentSheet Example/PaymentSheet Example/EndpointSelectorViewController.swift [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+struct EndpointSpec: Codable {
+    let endpointMap: [String: [String]]
+
+    private enum CodingKeys: String, CodingKey {
+        case endpointMap = "endpoint_map"
+    }
+    init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        self.endpointMap = try container.decode([String: [String]].self, forKey: .endpointMap)
+    }
```
Comment (eurias-stripe) [inline Example/PaymentSheet Example/PaymentSheet Example/EndpointSelectorViewController.swift:189]: You shouldn't need all these if you use `KeyDecodingStrategy.convertFromSnakeCase`

File: Example/PaymentSheet Example/PaymentSheet Example/PaymentSheet Example/PaymentSheetTestPlayground.swift [hunk 3]
```diff
@@ -249,6 +252,7 @@ class PaymentSheetTestPlayground: UIViewController {
     var clientSecret: String?
     var ephemeralKey: String?
     var customerID: String?
+    var checkoutEndpoint: String?
```
Comment (eurias-stripe) [inline Example/PaymentSheet Example/PaymentSheet Example/PaymentSheet Example/PaymentSheetTestPlayground.swift:255]: Make this a non optional and initialize it to the `defaultCheckoutEndpoint`

**Review by eurias-stripe — CHANGES_REQUESTED** [review]

File: Example/PaymentSheet Example/PaymentSheet Example/EndpointSelectorViewController.swift [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+        let sectionNumberIndex = sectionNumber - 1
+        guard let sortedSection = sortedSections() else {
+            return "Loading..."
+        }
+        return sortedSection[sectionNumberIndex].key
+    }
+
+    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = UITableViewCell(style: .default, reuseIdentifier: "endpoint")
```
[the remark this answers is not in the document]
Comment (wooj-stripe) [reply Example/PaymentSheet Example/PaymentSheet Example/EndpointSelectorViewController.swift:105]: Sorry, not sure what you are suggesting here?

File: Example/PaymentSheet Example/PaymentSheet Example/EndpointSelectorViewController.swift [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+            return nil
+        }
+        let sectionKey = sortedSection[indexPath.section-1].key
+        guard let endpointsInSection = specs.endpointMap[sectionKey] else {
+            return nil
+        }
+        return endpointsInSection[indexPath.row]
+    }
+
+    func sortedSections() -> [Dictionary<String, [String]>.Element]? {
```
[the remark this answers is not in the document]
Comment (wooj-stripe) [reply Example/PaymentSheet Example/PaymentSheet Example/EndpointSelectorViewController.swift]: This does not seem to compile.  Reverting.

File: Example/PaymentSheet Example/PaymentSheet Example/EndpointSelectorViewController.swift [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+        } catch {
+            return nil
+        }
+    }
+}
+
+// MARK: - Helpers
+extension EndpointSelectorViewController {
+    func endpointFor(indexPath: IndexPath) -> String? {
+        if indexPath.section == 0 {
```
[the remark this answers is not in the document]
Comment (wooj-stripe) [reply Example/PaymentSheet Example/PaymentSheet Example/EndpointSelectorViewController.swift:153]: I feel like this is way more readable, but sure. okay, changed to:
```
        guard indexPath.section != 0 else {
            return currentCheckoutEndpoint
        }
```

File: Example/PaymentSheet Example/PaymentSheet Example/PaymentSheet Example/PaymentSheetTestPlayground.swift (hunk 3 above)
Comment (wooj-stripe) [reply Example/PaymentSheet Example/PaymentSheet Example/PaymentSheet Example/PaymentSheetTestPlayground.swift:255]: Was hesitant about doing this at first but... Didn't. Thanks for calling this out!

File: Example/PaymentSheet Example/PaymentSheet Example/EndpointSelectorViewController.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (eurias-stripe) [reply Example/PaymentSheet Example/PaymentSheet Example/EndpointSelectorViewController.swift:105]:
```suggestion
        let cell = dequeueReusableCell(withIdentifier: "endpoint", for: indexPath)
```
And you need to have registered the cell type before

File: Example/PaymentSheet Example/PaymentSheet Example/EndpointSelectorViewController.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (eurias-stripe) [reply Example/PaymentSheet Example/PaymentSheet Example/EndpointSelectorViewController.swift:167]: I had an extra bracket 🙈

File: Example/PaymentSheet Example/PaymentSheet Example/PaymentSheet Example/PaymentSheetTestPlayground.swift [hunk 7]
```diff
@@ -15,6 +15,9 @@ import UIKit
 import SwiftUI
 import PassKit
 
+let endpointSelectorEndpoint = "https://stripe-mobile-payment-sheet-test-playground-v6.glitch.me/endpoints"
+let defaultCheckoutEndpoint = "https://stripe-mobile-payment-sheet-test-playground-v6.glitch.me/checkout"
+
```
Comment (eurias-stripe) [inline Example/PaymentSheet Example/PaymentSheet Example/PaymentSheet Example/PaymentSheetTestPlayground.swift:20]: These would be better inside `PaymentSheetTestPlayground`, make them `static` so you can access the default one before init

**Review by eurias-stripe — CHANGES_REQUESTED** [review]

File: Example/PaymentSheet Example/PaymentSheet Example/EndpointSelectorViewController.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (wooj-stripe) [reply Example/PaymentSheet Example/PaymentSheet Example/EndpointSelectorViewController.swift:105]: OH, right! it's been a while. oops.

File: Example/PaymentSheet Example/PaymentSheet Example/EndpointSelectorViewController.swift [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+        let sectionNumberIndex = sectionNumber - 1
+        guard let sortedSection = sortedSections() else {
+            return "Loading..."
+        }
+        return sortedSection[sectionNumberIndex].key
+    }
+
+    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let reuseIdentifier = "endpointCell"
+        let cell = tableView.dequeueReusableCell(withIdentifier: reuseIdentifier) ??
```
Comment (eurias-stripe) [inline Example/PaymentSheet Example/PaymentSheet Example/EndpointSelectorViewController.swift:106]: use the variant that takes the index path, that always returns a valid cell

File: Example/PaymentSheet Example/PaymentSheet Example/EndpointSelectorViewController.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (wooj-stripe) [reply Example/PaymentSheet Example/PaymentSheet Example/EndpointSelectorViewController.swift]: For whatever reason, this is not working.  I am reverting this, and hoping that this is just a "nit"

File: Example/PaymentSheet Example/PaymentSheet Example/EndpointSelectorViewController.swift (hunk 8 above)
Comment (wooj-stripe) [reply Example/PaymentSheet Example/PaymentSheet Example/EndpointSelectorViewController.swift:106]: I tried using:
```
        let cell = tableView.dequeueReusableCell(withIdentifier: reuseIdentifier, for: indexPath)
```
I'm seeing:
```
libc++abi: terminating with uncaught exception of type NSException
*** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'unable to dequeue a cell with identifier endpointCell - must register a nib or a class for the identifier or connect a prototype cell in a storyboard'
```

Comment (wooj-stripe) [reply Example/PaymentSheet Example/PaymentSheet Example/EndpointSelectorViewController.swift:106]: Screen Shot 2022-12-08 at 5 01 46 PM

Comment (wooj-stripe) [reply Example/PaymentSheet Example/PaymentSheet Example/EndpointSelectorViewController.swift]: Ah, ok. Added
```
self.tableView.register(UITableViewCell.self, forCellReuseIdentifier: reuseIdentifierEndpointCell)
```
in `viewDidLoad`

File: Example/PaymentSheet Example/PaymentSheet Example/EndpointSelectorViewController.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (eurias-stripe) [reply Example/PaymentSheet Example/PaymentSheet Example/EndpointSelectorViewController.swift:168]: It is just a nit, and yeah, it looks like this doesn't work as the return type of a function for some reason

**Review by eurias-stripe — APPROVED** [review]

File: Example/PaymentSheet Example/PaymentSheet Example/PaymentSheet Example/PaymentSheetTestPlayground.swift [hunk 9]
```diff
@@ -536,6 +550,20 @@ extension PaymentSheetTestPlayground: AddressViewControllerDelegate {
     }
 }
 
+// MARK: - EndpointSelectorViewControllerDelegate
+extension PaymentSheetTestPlayground: EndpointSelectorViewControllerDelegate {
+    func selected(endpoint: String) -> Void {
```
Comment (eurias-stripe) [inline Example/PaymentSheet Example/PaymentSheet Example/PaymentSheet Example/PaymentSheetTestPlayground.swift:555]: remove `-> Void`

**Pull request merged**